src/Shared/Dictionary/AdaptiveCapacityDictionary.cs

```diff
@@ -0,0 +1,712 @@
+                }
+
+                EnsureCapacity(Count + 1);
+
+                if (_dictionaryStorage != null)
+                {
+                    return _dictionaryStorage.TryAdd(key, value);
+                }
+
+                _arrayStorage[Count] = new KeyValuePair<TKey, TValue>(key, value);
+                _count++;
+                return true;
+            }
+
+            return _dictionaryStorage!.TryAdd(key, value);
+        }
+
+        /// <inheritdoc />
+        public bool TryGetValue(TKey key, [MaybeNullWhen(false)] out TValue value)
+        {
+            if (key == null)
+            {
+                ThrowArgumentNullExceptionForKey();
+            }
+
+            if (_arrayStorage != null)
+            {
+                return TryFindItem(key, out value);
+            }
+
+            return _dictionaryStorage!.TryGetValue(key, out value);
+        }
+
+        [DoesNotReturn]
+        private static void ThrowArgumentNullExceptionForKey()
+        {
+            throw new ArgumentNullException("key");
+        }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private void EnsureCapacity(int capacity)
+        {
+            EnsureCapacitySlow(capacity);
+        }
+
+        private void EnsureCapacitySlow(int capacity)
+        {
+            Debug.Assert(_arrayStorage != null);
+
+            if (_arrayStorage.Length < capacity)
+            {
+                if (capacity > DefaultArrayThreshold)
+                {
+                    _dictionaryStorage = new Dictionary<TKey, TValue>(capacity);
+                    foreach (var item in _arrayStorage)
+                    {
+                        _dictionaryStorage[item.Key] = item.Value;
+                    }
+
+                    // Clear array storage.
+                    _arrayStorage = null;
+                }
+                else
+                {
+                    capacity = _arrayStorage.Length == 0 ? DefaultArrayThreshold : _arrayStorage.Length * 2;
+                    var array = new KeyValuePair<TKey, TValue>[capacity];
+                    if (_count > 0)
+                    {
+                        Array.Copy(_arrayStorage, 0, array, 0, _count);
+                    }
+
+                    _arrayStorage = array;
+                }
+            }
+        }
+
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        private int FindIndex(TKey key)
+        {
+            // Generally the bounds checking here will be elided by the JIT because this will be called
+            // on the same code path as EnsureCapacity.
+            Debug.Assert(_dictionaryStorage == null);
+            Debug.Assert(_arrayStorage != null);
+
+            var array = _arrayStorage;
+            var count = _count;
+
+            for (var i = 0; i < count; i++)
+            {
+                if (_comparer.Equals(array[i].Key, key))
+                {
+                    return i;
+                }
+            }
```

Comment:
Same ways as `TryFindItem`:
```suggestion
            if (_count > 0)
            {
                for (var i = 0; i < ArrayStorageSpan.Length; ++i)
                {
                    if (_comparer.Equals(ArrayStorageSpan[i].Key, key))
                    {
                        return i;
                    }
                }
            }
```